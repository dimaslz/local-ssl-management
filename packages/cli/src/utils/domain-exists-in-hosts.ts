import fs from "fs/promises";

export async function domainExistsInHosts(domain: string) {
  const hostsContent = await fs.readFile("/etc/hosts", { encoding: "utf8" });

  const [alreadyExistsInHosts] =
    hostsContent.match(new RegExp(`\\d+.\\d+.\\d+.\\d+.*?${domain}$`, "mi")) ||
    [];

  return alreadyExistsInHosts;
}
