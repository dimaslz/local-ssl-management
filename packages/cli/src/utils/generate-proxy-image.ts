import type { Config } from "@dimaslz/local-ssl-management-core";
import { getLocalIP, mkcert } from "@dimaslz/local-ssl-management-core";
import Table from "cli-table3";
import consola from "consola";
import fs from "fs";
import path from "path";
import shell from "shelljs";

import listContainer from "@/list-container";
import Templates from "@/templates";

const LOCAL_IP = getLocalIP();
const distPath = path.resolve(__dirname, "./");
const rootPath = `${distPath}/.local-ssl-management`;
const sslPath = `${rootPath}/ssl`;

const renderTable = (config: Config[]) => {
  const tablePing = new Table({
    head: ["domain", "app running"],
  });

  config.forEach((c: Config) => {
    c.domain
      .split(",")
      .map((d) => d.trim())
      .forEach((domain) => {
        const curl = `curl -s -o /dev/null -w "%{http_code}" https://${domain}`;
        const status = shell.exec(curl, { silent: true }).stdout;

        if (status === "200") {
          tablePing.push([`https://${domain}`, "yes"]);
        } else {
          tablePing.push([`https://${domain}`, "no"]);
        }
      });
  });

  shell.echo(`\n${tablePing.toString()}\n`);
};

export const generateProxyImage = (config: Config[]) => {
  if (!config.length) {
    consola.warn("Does not exists config to create reverse proxy");

    return;
  }

  const localhostCertExists = fs.existsSync(sslPath + "/localhost-cert.pem");
  const localhostKeyExists = fs.existsSync(sslPath + "/localhost-key.pem");

  if (!localhostCertExists || !localhostKeyExists) {
    mkcert("localhost", sslPath);
  }

  const certsUrl: { cert: string; key: string }[] = [];

  const serverConfigs: string[] = config.map((c: Config) => {
    const nginxConfServerTpl: string = Templates.nginxConfServer || "";

    const multipleDomains = c.domain.split(",").length > 1;
    const domain = multipleDomains
      ? c.domain.split(",").join(" ")
      : c.domain.trim();
    const certName = multipleDomains
      ? c.domain.split(",").join("_")
      : c.domain.trim();

    const certFileExists = fs.existsSync(`${sslPath}/${certName}-cert.pem`);
    const keyFileExists = fs.existsSync(`${sslPath}/${certName}-key.pem`);
    const shouldCreateCerts = !certFileExists || !keyFileExists;

    if (shouldCreateCerts) {
      mkcert(domain, sslPath);
    }

    certsUrl.push({
      cert: `ssl/${certName}-cert.pem`,
      key: `ssl/${certName}-key.pem`,
    });

    const nginxLocationTpl: string = Templates.nginxLocation;

    // setup location
    let locationConfigs = c.services
      .map((service) => {
        return nginxLocationTpl
          .replace("%LOCATION%", service.location)
          .replace("%LOCAL_IP%", LOCAL_IP);
      })
      .join("\n\n");

    // setup ports
    c.services.forEach((service) => {
      locationConfigs = locationConfigs.replace("%PORT%", String(service.port));
    });

    return nginxConfServerTpl
      .replace(/%APP_DOMAIN%/g, certName)
      .replace(
        "%SERVER_NAME%",
        c.domain
          .split(",")
          .map((i) => i.trim())
          .join(" "),
      )
      .replace("#--server-location--#", locationConfigs);
  });

  const nginxConfTpl = Templates.nginxConf;
  const nginxConf = nginxConfTpl.replace(
    /\t\t#--server-config--#/,
    `\t\t${serverConfigs.join("\n")}`,
  );

  const nginxConfDest = `${rootPath}/nginx.conf`;
  fs.writeFileSync(nginxConfDest, nginxConf);

  const dockerfileContent = Templates.dockerfile;
  const dockerfileDest = `${rootPath}/Dockerfile`;
  fs.writeFileSync(
    dockerfileDest,
    dockerfileContent
      .replace(
        "#-certs-#",
        certsUrl
          .map((d: { cert: string; key: string }) => {
            return `COPY ${d.key} /etc/nginx/
COPY ${d.cert} /etc/nginx/`;
          })
          .join("\n\n"),
      )
      .replace(/#-main-path-#/g, ""),
  );

  shell.exec(
    `NAME=local-ssl-management && \
docker rm -f $NAME && \
docker rmi -f $NAME && \
docker build --no-cache -t $NAME ${rootPath} && \
docker run --name $NAME -p 80:80 -p 443:443 -d $NAME`,
    { silent: true },
  );

  listContainer();

  consola.success("SSL proxy running");

  renderTable(config);
};
