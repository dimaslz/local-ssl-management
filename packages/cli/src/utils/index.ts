export { default as validateDomain } from "./validate-domain";
export { default as validateLocation } from "./validate-location";
export { default as validatePort } from "./validate-port";
