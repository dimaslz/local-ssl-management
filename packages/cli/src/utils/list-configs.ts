import type { Config } from "@dimaslz/local-ssl-management-core";
import Table from "cli-table3";
import shell from "shelljs";

const isTest = process.env.NODE_ENV === "test";

export const listConfigs = (config: Config[]) => {
  const table = new Table({
    head: ["id", "key", "domains", "location", "port", "ping"],
    ...(isTest
      ? {
          style: {
            head: [], //disable colors in header cells
            border: [], //disable colors for the border
          },
        }
      : {}),
  });

  config.forEach(({ id, domain, services }) => {
    table.push([
      id,
      domain
        .split(",")
        .map((i) => i.trim())
        .join("_"),
      domain
        .split(",")
        .map((d) => `https://${d.trim()}`)
        .join(", "),
      "",
      "",
    ]);

    services.forEach((service) => {
      const curl = `curl -s -o /dev/null -w "%{http_code}" https://${domain}${service.location}`;
      const status = shell.exec(curl, { silent: true }).stdout;

      table.push([
        "",
        "",
        "",
        `${service.location}`,
        `${service.port}`,
        status,
      ]);
    });
  });

  shell.echo(`\n${table.toString()}\n`);
};
