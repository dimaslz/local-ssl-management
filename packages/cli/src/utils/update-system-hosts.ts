import { spawn } from "child_process";
import chokidar from "chokidar";
import consola from "consola";
import path from "path";
import prompt from "prompt";

const distPath = path.resolve(__dirname, "./");
const runPath = `${distPath}/run`;

async function promptPassword() {
  prompt.start();
  prompt.message = "";
  prompt.delimiter = ":";
  const { password } = await prompt.get({
    properties: {
      password: {
        message: "Password",
        hidden: true,
      },
    },
  });
  return password;
}

export function updateSystemHosts(tmpHostsPath: string) {
  const sudoCommand = spawn("sudo", [
    "-S",
    "-k",
    "node",
    `${runPath}/index.js`,
    tmpHostsPath,
  ]);

  sudoCommand.stderr.on("data", async () => {
    const password = await promptPassword();
    sudoCommand.stdin.write(password + "\n");
  });

  const watchTmpHostsFile = chokidar.watch(tmpHostsPath).on("unlink", () => {
    watchTmpHostsFile.unwatch(tmpHostsPath);

    consola.info(`Your hosts has been updated. Type \`cat/etc/hosts\` in your terminal to see the last updates on LOCAL SSL block.
`);
  });
}
