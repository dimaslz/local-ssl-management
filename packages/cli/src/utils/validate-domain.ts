import chalk from "chalk";
import isUrlHttp from "is-url-http";
import shell from "shelljs";

const validateDomain = (value: string) => {
	const domains = value.split(",").map(d => `https://${d.trim()}`);

	domains.forEach((domainItem) => {
		let domain = domainItem;

		// include validate domains like .local or .test
		if (/\.local$|\.test$/g.test(domainItem)) {
			domain = domain.replace(/\.local$|\.test$/, ".com");
		}

		if (!isUrlHttp(domain)) {
			shell.echo(chalk.red(`\n[Error] - Domain (${domainItem})format is not valid\n`));
			shell.exit(1);
		}
	});
};

export default validateDomain;