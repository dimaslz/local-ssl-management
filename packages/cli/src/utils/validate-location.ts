import consola from "consola";

export function validateLocation(location: string) {
  if (!location.startsWith("/")) {
    consola.error("Location should start by /");

    return;
  }
}
