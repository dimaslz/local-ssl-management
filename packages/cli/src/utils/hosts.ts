import consola from "consola";
import fs from "fs";
import path from "path";

import { HOSTS_END, HOSTS_START } from "@/constants";
import { updateSystemHosts } from "@/utils";

const distPath = path.resolve(__dirname, "./");
const tmpHostsPath = `${distPath}/.tmp-hosts`;

export async function getContentFromHosts(): Promise<string> {
  const hostsContent = await fs.readFileSync("/etc/hosts", {
    encoding: "utf8",
  });

  const [, localSSLHosts = ""] =
    hostsContent.match(new RegExp(`${HOSTS_START}([^#]+)${HOSTS_END}`, "im")) ||
    [];

  return localSSLHosts?.trim();
}

export async function setContentToHosts(content: string): Promise<string> {
  const hostsContent = await fs.readFileSync("/etc/hosts", {
    encoding: "utf8",
  });
  const [, localSSLHosts = ""] =
    hostsContent.match(new RegExp(`${HOSTS_START}([^#]+)${HOSTS_END}`, "im")) ||
    [];

  let newContent = "";
  if (!localSSLHosts) {
    newContent = `${hostsContent}\n\n${HOSTS_START}\n${content}\n${HOSTS_END}\n`;
  } else {
    newContent = hostsContent.replace(
      new RegExp(`${HOSTS_START}[^#]+${HOSTS_END}`, "im"),
      `${HOSTS_START}\n${content}\n${HOSTS_END}`,
    );
  }

  consola.info(`The following peace of text is what you need to update in your \`/etc/hosts\`. You can do it manually or, if you want to do it now, type your password to have permission to update the hosts.

${HOSTS_START}\n${content}\n${HOSTS_END}
`);

  const answer = await consola.prompt("Do you want to update your hosts now?", {
    type: "confirm",
  });

  if (answer) {
    await fs.writeFileSync(tmpHostsPath, newContent);
    updateSystemHosts(tmpHostsPath);
  }

  return newContent;
}

export async function updateHosts(domain: string) {
  const hostsContent = await fs.readFileSync("/etc/hosts", {
    encoding: "utf8",
  });
  const currentLocalSSLHosts = await getContentFromHosts();

  const [alreadyExistsInLocalSSLHosts] =
    currentLocalSSLHosts.match(
      new RegExp(`\\d+.\\d+.\\d+.\\d+.*?${domain}$`, "mi"),
    ) || [];

  if (alreadyExistsInLocalSSLHosts) {
    return;
  }

  const [alreadyExists] =
    hostsContent.match(new RegExp(`\\d+.\\d+.\\d+.\\d+.*?${domain}$`, "mi")) ||
    [];

  let hosts = "";
  if (alreadyExists) {
    const hostsContentCleaned = hostsContent.replace(
      new RegExp(`\\d+.\\d+.\\d+.\\d+.*?${domain}$`, "mi"),
      "",
    );

    try {
      await fs.writeFileSync(tmpHostsPath, `\n${hostsContentCleaned.trim()}\n`);
      updateSystemHosts(tmpHostsPath);
    } catch (error) {
      // silent
    }
    hosts = `${currentLocalSSLHosts}\n${alreadyExists.trim()}`;
  } else {
    const [alreadyExists] =
      currentLocalSSLHosts.match(
        new RegExp(`\\d+.\\d+.\\d+.\\d+.*?${domain}$`, "mi"),
      ) || [];
    if (!alreadyExists) {
      hosts = `${currentLocalSSLHosts}\n127.0.0.1               ${domain}`;
    }
  }

  return await setContentToHosts(hosts);
}
