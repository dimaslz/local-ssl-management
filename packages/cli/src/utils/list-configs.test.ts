import shell from "shelljs";

import { listConfigs } from "@/utils/list-configs";

describe("Utils - listConfigs", () => {
  test("list configs", () => {
    const configs = [
      {
        id: "d7274462-e2cc-4ed3-8b07-fea3e342e7aa",
        domain: "local.some-domain.tld",
        services: [
          {
            id: "29ecf855-5262-4b53-b742-03c8505abf2f",
            location: "/",
            port: "3333",
          },
          {
            id: "3cbf00b7-4fdd-4d37-8634-575c969faa15",
            location: "/app-name",
            port: "3000",
          },
        ],
      },
    ];

    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    (vi.spyOn(shell, "exec") as any).mockImplementation(() => ({
      stdout: 200,
    }));

    listConfigs(configs);

    expect(shell.echo).matchSnapshot();
  });
});
