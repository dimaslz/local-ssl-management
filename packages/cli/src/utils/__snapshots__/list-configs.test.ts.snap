// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Utils - listConfigs > list configs 1`] = `
[MockFunction spy] {
  "calls": [
    [
      "
┌──────────────────────────────────────┬───────────────────────┬───────────────────────────────┬───────────┬──────┬──────┐
│ id                                   │ key                   │ domains                       │ location  │ port │ ping │
├──────────────────────────────────────┼───────────────────────┼───────────────────────────────┼───────────┼──────┼──────┤
│ d7274462-e2cc-4ed3-8b07-fea3e342e7aa │ local.some-domain.tld │ https://local.some-domain.tld │           │      │      │
├──────────────────────────────────────┼───────────────────────┼───────────────────────────────┼───────────┼──────┼──────┤
│                                      │                       │                               │ /         │ 3333 │ 200  │
├──────────────────────────────────────┼───────────────────────┼───────────────────────────────┼───────────┼──────┼──────┤
│                                      │                       │                               │ /app-name │ 3000 │ 200  │
└──────────────────────────────────────┴───────────────────────┴───────────────────────────────┴───────────┴──────┴──────┘
",
    ],
  ],
  "results": [
    {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;
